libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,198 @@ katana::EdgeTypeAwareTopology::MakeFrom(
       pg, edge_type_index, e_topo, std::move(per_type_adj_indices)});
 }
 
+std::unique_ptr<katana::ProjectedTopology>
+katana::ProjectedTopology::MakeTypeProjectedTopology(
+    const katana::PropertyGraph* pg, const std::vector<std::string>& node_types,
+    const std::vector<std::string>& edge_types) {
+  KATANA_LOG_DEBUG_ASSERT(pg);
+
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedTopology pst;
+    return std::make_unique<ProjectedTopology>(std::move(pst));
+  }
+
+  // calculate number of new nodes
+  uint32_t num_new_nodes = 0;
+  uint32_t num_new_edges = 0;
+
+  katana::DynamicBitset bitset_nodes;
+  bitset_nodes.resize(topology.num_nodes());
+
+  NUMAArray<Node> original_to_projected_nodes_mapping;
+  original_to_projected_nodes_mapping.allocateInterleaved(topology.num_nodes());
+
+  if (node_types.empty()) {
+    num_new_nodes = topology.num_nodes();
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+      original_to_projected_nodes_mapping[src] = 1;
+    });
+  } else {
+    katana::ParallelSTL::fill(
+        original_to_projected_nodes_mapping.begin(),
+        original_to_projected_nodes_mapping.end(), Node{0});
+
+    auto node_entity_type_manager = pg->GetNodeTypeManager();
+
+    auto res_node_type =
+        node_entity_type_manager.GetNonAtomicEntityTypeFromStrings(node_types);
```

Comment:
There's a bug here (ditto for edges): 

There is no guarantee that the vector of node types has a corresponding entity type ID. 
(1) Consider "Person" and "Message". No node has both those types, so there will no entity type ID. 
(2) Consider "XYZ". No node may have this type, so there will be no entity type ID. 

This (caller) code should check if an error is returned. An error means that no node has this type, so you can skip the code below and create a graph with no nodes and edges (if an error is returned for edge types below, create a graph with only the selected nodes and no edges).